### PR TITLE
Support *args in Pipe.forward API.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57325 Add `NoChunk` wrapper for pipeline args.
* #57226 Support non-Tensor args in the Pipe API
* **#55441 Support *args in Pipe.forward API.**

This is the first step towards supporting the proposal outlined in
https://github.com/pytorch/pytorch/issues/53952.

In this PR I've ensured Pipe.forward() accepts a *inputs argument instead of
just a single input as previously. This lays the groundwork for supporting
non-Tensors and generic arguments to the Pipe API. In this PR we still only
support Tensors and non-Tensor support will come in future PRs.

For backward compatibility I've ensured a single Tuple[Tensor] input still
works as expected previously.

Differential Revision: [D27613887](https://our.internmc.facebook.com/intern/diff/D27613887/)

### CODE REVIEW

File: torch/distributed/pipeline/sync/microbatch.py [hunk 1]
```diff
@@ -132,39 +141,44 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
...
         TypeError: input is not a tensor or tensors.
 
     """
-    if isinstance(input, Sequence):
-        for x in input:
-            if not isinstance(x, Tensor):
-                raise TypeError(f"expected Tensor, but got {input.__class__.__name__}")
-        return
+    for input in inputs:
+        if isinstance(input, Sequence):
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/microbatch.py:152]: We allow Sequence only if it `len(inputs)==1` , right?

File: torch/distributed/pipeline/sync/pipe.py [hunk 2]
```diff
@@ -248,8 +240,9 @@ def __init__(
         if deferred_batch_norm:
             module = DeferredBatchNorm.convert_deferred_batch_norm(module, chunks)
 
-        self.partitions, self.devices = _split_module(module)
-        _verify_splitting(module, self.partitions, self.devices)
+        self.partitions = module
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/pipe.py:243]: could you elaborate how this change helps what you wanted to achieve in this PR? with _split_module, we had the flexibility of changing partitioning algorithm by modifying _split_module implementation.

Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/pipe.py:243]: `split_module` was mostly redundant since we don't really split the module and in this API the module is already split by the user. Since this is an internal implementation detail, we can always change it later if needed.

File: torch/distributed/pipeline/sync/pipeline.py [hunk 3]
```diff
@@ -82,7 +82,7 @@ class Pipeline:
 
     def __init__(
         self,
-        partitions: List[nn.Sequential],
+        partitions: nn.Sequential,
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/pipeline.py:85]: partitions seems to be list not a module

File: torch/distributed/pipeline/sync/pipe.py [hunk 4]
```diff
@@ -308,7 +329,10 @@ def __getitem__(self, index: int) -> nn.Module:
     def __iter__(self) -> Iterable[nn.Module]:
         """Iterates over children of the underlying sequential module."""
         for partition in self.partitions:
-            yield from partition
+            if isinstance(partition, nn.Sequential):
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/pipe.py:332]: is this change needed?

File: torch/distributed/pipeline/sync/pipe.py [hunk 5]
```diff
@@ -291,12 +307,17 @@ def __getitem__(self, index: int) -> nn.Module:
             partitions = partitions[::-1]
 
         for partition in partitions:
-            try:
-                return partition[index]
-            except IndexError:
-                pass
 
-            shift = len(partition)
+            if isinstance(partition, nn.Sequential):
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/pipe.py:311]: are these changes needed?

File: torch/distributed/pipeline/sync/pipe.py [hunk 6]
```diff
@@ -282,7 +298,7 @@ def __init__(
 
     def __len__(self) -> int:
         """Counts the length of the underlying sequential module."""
-        return sum(len(p) for p in self.partitions)
+        return sum(len(p) if isinstance(p, nn.Sequential) else 1 for p in self.partitions)
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/pipe.py:301]: revert this?

File: torch/distributed/pipeline/sync/pipe.py [hunk 7]
```diff
@@ -68,7 +68,7 @@ def _verify_module(module: nn.Sequential) -> None:
 
 
 def _verify_splitting(
-    module: nn.Sequential, partitions: List[nn.Sequential], devices: List[torch.device]
+    module: nn.Sequential, partitions: nn.Sequential, devices: List[torch.device]
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/pipe.py:71]: revert this?

File: torch/distributed/pipeline/sync/microbatch.py [hunk 8]
```diff
@@ -40,9 +40,11 @@ class Batch:
 
     """
 
-    def __init__(self, value: TensorOrTensors) -> None:
+    def __init__(self, value: TensorOrTensors, is_single_sequence=False) -> None:
         self.value = value
         self.atomic = torch.is_tensor(value)
+        # For backward compatibility.
+        self.is_single_sequence = is_single_sequence
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/microbatch.py:47]: how `is_single_sequence` is different from `atomic`?

Comment (mrzzd) [reply torch/distributed/pipeline/sync/microbatch.py:47]: Sounds good. Maybe it would be better if you add a description of `is_single_sequence` to the docstring

File: test/distributed/pipeline/sync/test_microbatch.py [hunk 9]
```diff
@@ -44,7 +44,7 @@ def test_batch_non_atomic():
 
 def test_batch_call():
     a = Batch(torch.tensor(42))
-    b = Batch((torch.tensor(42), torch.tensor(21)))
+    b = Batch((torch.tensor(42), torch.tensor(21)), is_single_sequence=True)
```
Comment (mrzzd) [inline test/distributed/pipeline/sync/test_microbatch.py:47]: it seems we do not have a similar test with is_single_sequence=False (that tests .call method) here.

File: torch/distributed/pipeline/sync/microbatch.py [hunk 10]
```diff
@@ -132,39 +141,44 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
...
+        for tensor in inputs:
             tensors = tensor.chunk(chunks)
-            rotated.append(cast(Tensors, tensors))
+            rotated.append(tensors)
 
-        inputs = zip(*rotated)
+        unwrapped_inputs = zip(*rotated)  # type: ignore[assignment]
 
-    return [Batch(x) for x in inputs]
+    return [Batch(x, is_single_sequence) for x in unwrapped_inputs]
```
Comment (mrzzd) [inline torch/distributed/pipeline/sync/microbatch.py:181]: Suppose the user creates `Pipe` for `nn.Sequence(m1: M1, m2: M2)` with following signature:
```py
class M1(nn.Module):
   def forward(self, input: Tensor) ->  Tuple[Tensor, Tensor]:
        ...
class M2(nn.Module):
   def forward(self, input: Tuple[Tensor, Tensor]) ->  Tensor:
        ...
```
This works fine with old implementation, but it seems to me with new implementation we set is_single_sequence=False, which is wrong.

Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/microbatch.py]: @mrzzd I thought about this for a while and couldn't come up with a good solution. I did consider your proposal about adding `is_single_sequence` to the constructor. The tricky thing with that is that if by default that is False, we anyways end up breaking compatibility since the user has to manually set this flag to get the old behavior.

I felt it was probably better to just break compatibility since the feature is Beta and keep the behavior simple and consistent without special handling for tuple of tensors. We can call this out in our release notes as well for the next release. As a result, I removed most of the logic pertaining to is_single_sequence.

cc @mrshenli

File: torch/distributed/pipeline/sync/microbatch.py [hunk 11]
```diff
@@ -132,39 +135,32 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
         self.value = value[0]
 
 
-def check(input: TensorOrTensors) -> None:
+def check(*inputs) -> None:
```
Comment (SciPioneer) [inline torch/distributed/pipeline/sync/microbatch.py:138]: nit: no type hint?

File: torch/distributed/pipeline/sync/microbatch.py [hunk 12]
```diff
@@ -132,39 +135,32 @@ def _setitem_by_slice(self, index: slice, value: Tensors) -> None:
...
+    for input in inputs:
+        if not isinstance(input, Tensor):
+            raise TypeError(f"expected Tensor, but got {input.__class__.__name__}")
 
-    if not isinstance(input, Tensor):
-        raise TypeError(f"expected Tensor, but got {input.__class__.__name__}")
 
-
-def scatter(input: TensorOrTensors, chunks: int) -> List[Batch]:
+def scatter(*inputs, chunks: int) -> List[Batch]:
```
Comment (SciPioneer) [inline torch/distributed/pipeline/sync/microbatch.py:150]: nit: no type hint?

**Review by wayi1 — APPROVED** [review]
LGTM, only one question about the type hint.

File: test/distributed/pipeline/sync/test_pipe.py [hunk 13]
```diff
@@ -650,3 +636,19 @@ def forward(self, x):
...
+def test_multiple_inputs(checkpoint, setup_rpc):
+    class Module1(nn.Module):
+        def forward(self, a, b, c):
+            return a + b + c, a * b * c
+
+    class Module2(nn.Module):
+        def forward(self, a, b):
+            return a + b
+
+    model = Pipe(nn.Sequential(Module1().cuda(0), Module2().cuda(0)), chunks=2, checkpoint=checkpoint)
```
Comment (mrzzd) [inline test/distributed/pipeline/sync/test_pipe.py:651]: should change to PipeSequential

**Review by mrzzd — APPROVED** [review]
Thanks Pritam.

File: test/distributed/pipeline/sync/test_pipe.py (hunk 13 above)
Comment (pritamdamania87) [reply test/distributed/pipeline/sync/test_pipe.py:651]: PipeSequential should only be used internally inside the Pipe API right? I would not expect users to use PipeSequential.

File: torch/distributed/pipeline/sync/microbatch.py (hunk 11 above)
Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/microbatch.py:138]: The PRs on top allow for generic inputs, as a result I didn't put a type hint. This is similar to *args.

File: torch/distributed/pipeline/sync/microbatch.py (hunk 12 above)
Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/microbatch.py:150]: Same as above.
